### PR TITLE
Add an audit system to Spack

### PR DESCRIPTION
**Pull request by alalazo**
This PR adds a way to check various parts of Spack from the command line. The motivation to start this has been seeing a few issues where errors / inconsistencies in package recipes where mistaken for bugs in the new concretizer, due to poor reporting, and a debug session with @jjellio  in which after much struggle we understood that an issue was due to a compiler being defined multiple times in the configuration (which is not allowed but Spack was not clear at all on what was the problem).

The PR should introduce:
- [x] A way to declare new class of audits (e.g. audits for compiler configuration, audits for package directives etc.)
- [x] A way to register functions as part of those audits 
- [x] A command to drive this from the terminal
- [x] Documentation of the new feature
- [x] Tests for the `spack.audit` module 

As a first example, recreating a case like the one debugged with @jjellio we can obtain with this PR:
```console
culpo@MacBook-Pro bootstrap % cat spack.yaml             
spack:
  specs:
  - zlib
  compilers:
  - compiler:
      spec: gcc@10.2.0
      paths:
        cc: /Users/culpo/PycharmProjects/spack/opt/spack/darwin-bigsur-cannonlake/apple-clang-12.0.0/gcc-10.2.0-zm6u6otbxbacpvsbx6lugzo6qeb2hd7v/bin/gcc
        cxx: /Users/culpo/PycharmProjects/spack/opt/spack/darwin-bigsur-cannonlake/apple-clang-12.0.0/gcc-10.2.0-zm6u6otbxbacpvsbx6lugzo6qeb2hd7v/bin/g++
        f77: /Users/culpo/PycharmProjects/spack/opt/spack/darwin-bigsur-cannonlake/apple-clang-12.0.0/gcc-10.2.0-zm6u6otbxbacpvsbx6lugzo6qeb2hd7v/bin/gfortran
        fc: /Users/culpo/PycharmProjects/spack/opt/spack/darwin-bigsur-cannonlake/apple-clang-12.0.0/gcc-10.2.0-zm6u6otbxbacpvsbx6lugzo6qeb2hd7v/bin/gfortran
      flags:
        cflags: '-O3'
      operating_system: bigsur
      target: x86_64
      modules: []
      environment: {}
      extra_rpaths: []
culpo@MacBook-Pro bootstrap % spack audit configuration     
culpo@MacBook-Pro bootstrap % spack -e . audit configuration
CFG-COMPILER: 1 issue found
1. Compiler defined multiple times: gcc@10.2.0
    in "/Users/culpo/spack/bootstrap/spack.yaml", line 5, column 5
    in "/Users/culpo/.spack/darwin/compilers.yaml", line 2, column 3
```

### CODE REVIEW

File: lib/spack/docs/basic_usage.rst
Comment (scheibelp) [inline lib/spack/docs/basic_usage.rst]: I think a bit more explanation on the intent of this would be useful. For example: why doesn't Spack just automatically check for these problems? Why don't they raise an exception and direct the user to do something?

In the duplicate compilers case, that was originally forbidden (and checked in the code) but then later relaxed (for reasons I cannot recall), so in that case we just stopped considering it a bug.

In the case of unknown variants in directives, I think the reason is that it would be expensive to perform this check every time there was an error (so we either have to make it manual like this or conditionally check it for a certain subset of our exceptions).

File: lib/spack/spack/test/cmd/audit.py
Comment (scheibelp) [inline lib/spack/spack/test/cmd/audit.py]: IMO this isn't verifying the correctness of the `audit` logic, but rather using the `audit` logic to verify the correctness of the packages. To that end, these two tests should exist in `package_sanity`.

For example to actually test `spack audit configs`, we'd want to create a failing config, then audit that and make sure the audit fails. I won't immediately demand that here but I'm going to think on whether that should be required for this PR.

File: lib/spack/spack/cmd/audit.py
Comment (scheibelp) [inline lib/spack/spack/cmd/audit.py]: Would it make sense to define the grouping in `AuditClass` as well as the tag? It seems strange to define the grouping here, especially if users can add their own checks in the future.

File: lib/spack/spack/cmd/audit.py
Comment (scheibelp) [inline lib/spack/spack/cmd/audit.py]: IMO this should be called `names` or `packages`. Also the typical multi-value arguments we process either specify a " "-delimited string in quotes or a ","-delimited string (I think usually the latter). If I'm missing some commands that operate the way you did it here, let me know. I think there is value in being consistent in how we expect users to supply multi-value args.

File: lib/spack/spack/audit.py
Comment (scheibelp) [inline lib/spack/spack/audit.py:73]: (high-level issue) This process feels more complex than it has to be. If it is designed to allow non-core devs to add audits, then why not use a similar mechanism to hooks? If not, then why use a registration mechanism like this (the `spack audit` command code could just call a function like `_search_duplicate_compilers` directly.

**Review by scheibelp — COMMENTED** [review]
I have a few preliminary comments but I think the most important discussion to address first is the one I marked as "high-level": I'm wondering about the necessity of the `AuditClass` class.

File: lib/spack/spack/audit.py
Comment (alalazo) [reply lib/spack/spack/audit.py:73]: The process is:
1. Generate a decorator if there's none that already fit 
2. Write a function that performs the check and decorate it

I wrote this to be able to group related checks under the same tag and have already some structure to extend it later via hooks. I was thinking of user code similar to:
```python
import spack.audit

@spack.audit.audit_cfgcmp
def _mycheck(error_cls):
    ...
```
that gets imported via hooks. Later the functions can be executed via `spack.audit.CALLBACKS` in a script or via the `spack audit` command from the command line. I'm not adding the hook in this PR since it is already big enough, and usually big PRs mean longer turnaround for review.

File: lib/spack/docs/basic_usage.rst
Comment (alalazo) [reply lib/spack/docs/basic_usage.rst]: I expanded a bit more the introduction to this section. Let me know what you think.

File: lib/spack/spack/test/cmd/audit.py
Comment (alalazo) [reply lib/spack/spack/test/cmd/audit.py]: I'm fine with doing that, I didn't add it in the first place since it requires some boilerplate to setup the test correctly. We should probably check that a configuration passing all sanity checks returns an error code of zero.

File: lib/spack/spack/cmd/audit.py
Comment (alalazo) [reply lib/spack/spack/cmd/audit.py]: I went with positional arguments:
```console
$ spack audit packages hdf5 zlib mpich
```
Let me know if this would be acceptable.

File: lib/spack/spack/audit.py
Comment (scheibelp) [reply lib/spack/spack/audit.py:73]: To confirm, I think we agreed that 

> Generate a decorator if there's none that already fit

Would be done by Spack developers (and we assume that the decorators defined in this PR will be generally sufficient). In that case I think this is sufficiently simple. Generally I don't think users should have to define decorators. 

I still think it might be worthwhile to create a "catch-all" decorator for audits that don't expect any arguments (they can still verify things by accessing global state for example).

File: lib/spack/spack/cmd/audit.py
Comment (scheibelp) [reply lib/spack/spack/cmd/audit.py]: I still think it makes sense to define the group when creating the decorator:

```
audit_pkgdirectives = AuditClass(
    tag='PKG-DIRECTIVES',
    group='packages',  # <--- define the group here
    description='Sanity checks on specs used in directives',
    kwargs=('pkgs',)
)
```

Comment (scheibelp) [reply lib/spack/spack/cmd/audit.py]: It probably still makes sense to keep the overall group names (`packages`, `configs`) here since each corresponds to specific subcommand argument processing logic. But I think the association of individual check to group should be done where the `AuditClass` is defined.

File: lib/spack/spack/cmd/audit.py
Comment (scheibelp) [reply lib/spack/spack/cmd/audit.py]: That sounds good to me: several of our commands specify arguments in this manner (e.g. `spack mirror create`).

File: lib/spack/spack/audit.py
Comment (alalazo) [reply lib/spack/spack/audit.py:73]: Done in 6482eeb

File: lib/spack/spack/cmd/audit.py
Comment (alalazo) [reply lib/spack/spack/cmd/audit.py]: Done in 9739b42

File: lib/spack/docs/basic_usage.rst
Comment (alalazo) [reply lib/spack/docs/basic_usage.rst]: This should be in b7d30a5

**Comment by alalazo** [comment]
@scheibelp Ready for a second review

File: lib/spack/docs/basic_usage.rst
Comment (scheibelp) [inline lib/spack/docs/basic_usage.rst]: I think it would be good to introduce `spack audit` command as quickly as possible like:


> The "spack audit" command can be used to detect a number of configuration issues. This command detects configuration settings which are not strictly wrong but are not likely to be useful outside of special cases.
>
> "spack audit" can also be used to detect dependency issues with packages - for example cases where a package constrains a dependency with a variant that doesn't exist (in this case Spack *could* report the problem ahead of time but automatically performing the check would slow down most runs of Spack).

File: lib/spack/spack/test/cmd/audit.py
Comment (scheibelp) [reply lib/spack/spack/test/cmd/audit.py]: These tests should be moved.

Examples of tests that would be good here:

* Create a package that specifies a "wrong" variant on a dependency and audit it
* Create a "bad" config and audit it

I think we should be testing the audits we provide

File: lib/spack/spack/audit.py
Comment (scheibelp) [inline lib/spack/spack/audit.py:157]: If I understand correctly, this won't be useful until we have logic similar to hooks as described in https://github.com/spack/spack/pull/23053/files#r630265011 - IMO it would be good to mention that in this comment.

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a couple minor requests for documentation and another request about adding/reorganizing tests.

File: lib/spack/docs/basic_usage.rst
Comment (alalazo) [reply lib/spack/docs/basic_usage.rst]: Should be addressed by add28d9

File: lib/spack/spack/audit.py
Comment (alalazo) [reply lib/spack/spack/audit.py:157]: Comment added in 56b3136

File: lib/spack/spack/test/cmd/audit.py
Comment (alalazo) [reply lib/spack/spack/test/cmd/audit.py]: Tests added in 48d6eaf

**Comment by alalazo** [comment]
@scheibelp I think I addressed the points above. Let me know if there are other changes needed.

File: lib/spack/spack/test/audit.py
Comment (scheibelp) [inline lib/spack/spack/test/audit.py]: Could we take a non-parameterization approach here? I'm getting lost reading 3-valued tuple spanning 30 lines.

File: lib/spack/spack/test/cmd/audit.py
Comment (scheibelp) [inline lib/spack/spack/test/cmd/audit.py]: As is, these tests are a bit "broad", I think it would be sensible to audit specific packages and expect a failure in some cases and a success in others.

**Review by scheibelp — CHANGES_REQUESTED** [review]
The new tests look great, although I have a request for simplifying one.

Also, should we be running these audit commands as part of `package_sanity`? I think it would definitely make sense to run the package audits. Perhaps we could hold off on that for a later time though.

File: lib/spack/spack/test/audit.py
Comment (alalazo) [reply lib/spack/spack/test/audit.py]: I just checked in a text editor and I concur with you. I am used to see these values within an IDE where they look like:
Screenshot from 2021-06-15 12-39-04

Instead of avoiding parametrization (since the test will be exactly the same) I'll extract the data and bind it to a private variable.

Comment (alalazo) [reply lib/spack/spack/test/audit.py]: Should be addressed by 16f38b6

File: lib/spack/spack/test/cmd/audit.py
Comment (alalazo) [reply lib/spack/spack/test/cmd/audit.py]: Should be addressed by 29551c2

**Comment by alalazo** [comment]
> Also, should we be running these audit commands as part of package_sanity? I think it would definitely make sense to run the package audits. Perhaps we could hold off on that for a later time though.

I thought about that and I think we should do it at some point, but as you said I'd hold off for a later PR.

Right now there are 76 warnings for the built-in repository. Most of them come from Paraview (see #23257 as an example of where they come from). I think a PR introducing this check in unit tests should also ensure that the built-in repository is clean of issues and that may require some iteration with package maintainers.

**Review by scheibelp — DISMISSED** [review]

**Comment by scheibelp** [comment]
Close/reopen to restart CI

**Comment by alalazo** [comment]
@scheibelp Pushed again to fix a conflict with command completion

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
Close/reopen to restart CI

**Comment by alalazo** [comment]
Thanks @scheibelp !

**Pull request merged**
